src/Rings/mpoly-graded.jl

```diff
@@ -1496,6 +1501,157 @@ function homogenization(f::MPolyElem, S::MPolyRing_dec, pos::Int = 1)
+   W = matrix(ZZ, W)
+   return _homogenization(f, W, var, pos)
+end
+
+@doc Markdown.doc"""
+    homogenization(f::MPolyElem, W::Union{fmpz_mat, Matrix{<:IntegerUnion}}, var::String, pos::Int = 1)
+
+If $m$ is the number of rows of `W`, extend the parent polynomial ring of `f` by inserting $m$ extra variables, starting at position `pos`.
+Correspondingly, extend the integer matrix `W` by inserting the standard unit vectors of size $m$ as new columns, starting at column `pos`.
+Grade the extended ring by converting the columns of the extended matrix to elements of the group $\mathbb Z^m$ and assigning these
+as weights to the variables. Homogenize `f` with respect to the induced $\mathbb Z^m$-grading on the original ring, using the extra variables as homogenizing
+variables. Return the result as an element of the extended ring with its $\mathbb Z^m$-grading. If $m=1$, the extra variable prints as `var`. Otherwise,
+the extra variables print as `var[`$i$`]`, for $i = 1 \dots m$.
+
+    homogenization(V::Vector{T},  W::Union{fmpz_mat, Matrix{<:IntegerUnion}}, var::String, pos::Int = 1) where {T <: MPolyElem}
+
+Given a vector `V` of elements in a common polynomial ring, create an extended ring with $\mathbb Z^m$-grading as above.
+Homogenize the elements of `V` correspondingly,  and return the vector of homogenized elements.
+
+    homogenization(I::MPolyIdeal{T},  W::Union{fmpz_mat, Matrix{<:IntegerUnion}}, var::String, pos::Int = 1; ordering::Symbol = :degrevlex) where {T <: MPolyElem}
+
+Return the homogenization of `I` in an extended ring with $\mathbb Z^m$-grading as above.
+
+!!! note
+    Applied to an ideal `I`, the function first homogenizes the generators of `I` in the extended ring.
+    It then creates the ideal generated by these homogenizations, and saturates this ideal
+    with respect to the ideal which is generated by the homogenizing variables.
+
+# Examples
+```jldoctest
+julia> R, (x, y) = PolynomialRing(QQ, ["x", "y"])
+(Multivariate Polynomial Ring in x, y over Rational Field, fmpq_mpoly[x, y])
+
+julia> f = x^3+x^2*y+x*y^2+y^3
+x^3 + x^2*y + x*y^2 + y^3
+
+julia> W = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> F = homogenization(f, W, "z", 3)
+x^3*z[1]^3*z[2]^3 + x^2*y*z[1]^2*z[2]^2 + x*y^2*z[1]*z[2] + y^3
+
+julia> parent(F)
+Multivariate Polynomial Ring in x, y, z[1], z[2] over Rational Field graded by
+  x -> [1 3]
+  y -> [2 4]
+  z[1] -> [1 0]
+  z[2] -> [0 1]
+```
+"""
+function homogenization(f::MPolyElem, W::Union{fmpz_mat, Matrix{<:IntegerUnion}}, var::String, pos::Int = 1)
+   return _homogenization(f, W, var, pos)
+end
+
+function homogenization(V::Vector{T},  W::Union{fmpz_mat, Matrix{<:IntegerUnion}}, var::String, pos::Int = 1) where {T <: MPolyElem}
+  @assert all(x->parent(x) == parent(V[1]), V)
+  R = parent(V[1])
+  A = String.(symbols(R))
+  l = length(A)
+  if (pos > l+1) ||  (pos <1)
+      throw(ArgumentError("Index out of range."))
+  end
```

Comment:
More idiomatic might be:
```suggestion
  pos in 1:l+1 || throw(ArgumentError("Index out of range."))
```